### PR TITLE
[model] feat: DeepSeek V4 QAT fake quant training

### PR DESCRIPTION
**Pull request by wuxibin89**
### What does this PR do?

| 模块 | 名称 | 参数 | 激活 | Block（参数 / 激活） | 推理侧计算 | 说明 |
|---|---|---|---|---|---|---|
| mHC | hc_* | FP32 | — | — | FP32 | 不量化 |
| Attn | wq_a / wq_b | FP8 | FP8 | 128×128 / 1×128 | 真 FP8 GEMM | |
| Attn | wkv | FP8 | FP8 | 128×128 / 1×128 | 真 FP8 GEMM | |
| Attn | **q（wq_b 输出）** | — | **不量化** | — | BF16 | 主 attention 的 Q 全程 BF16 |
| Attn | kv（SWA live） | — | FP8 | — / 1×64 | **BF16，仅存储** | 只量化 NoPE 448 维 |
| Attn | compressor 输出 KV | — | FP8 | — / 1×64 | **BF16，仅存储** | head=512，只量化 NoPE 448 维 |
| Attn | compressor.wkv/wgate | BF16 | — | — | FP32 | 不量化 |
| Attn | wo_a | FP8 | FP8 | 128×128 / 1×128 沿 r | 真 FP8 GEMM | 先 inv-RoPE 再整体量化，**含 RoPE**；参考实现是 BF16 |
| Attn | wo_b | FP8 | FP8 | 128×128 / 1×128 | 真 FP8 GEMM | |
| Indexer | wq_b | FP8 | FP8 | 128×128 / 1×128 | 真 FP8 GEMM | GEMM 输入侧 |
| Indexer | q（输出） | — | FP8 | — / 1×128 | 真 FP8×FP8 logits | 整个 head=128，含 RoPE，SM90 无 Hadamard |
| Indexer | compressor 输出 K | — | FP8 | — / 1×128 | 真 FP8×FP8 logits | 整个 head=128，含 RoPE |
| Indexer | weights_proj | BF16 | — | — | BF16 | 不量化 |
| MoE | gate | BF16 | — | — | FP32 | 不量化 |
| MoE | experts w1/w2/w3 | FP8 | FP8 | 128×128 / 1×128 | 真 FP8 GEMM | Flash 换 FP4：权重 `[out,in/32]`，激活仍 FP8 1×128 |
| MoE | shared_experts w1/w2/w3 | FP8 | FP8 | 128×128 / 1×128 | 真 FP8 GEMM | 各 checkpoint 一律 FP8 |

### Checklist Before Starting

- Search for relative PRs/issues and link here: ...
- PR title follows `[{modules}] {type}: {description}` format (enforced by [check_pr_title.yml](.github/workflows/check_pr_title.yml))
  - **Allowed modules:** `agent`, `ci`, `ckpt`, `config`, `data`, `dist`, `docker`, `docs`, `logging`, `lora`, `misc`, `model`, `omni`, `optim`, `ops`, `parallel`, `perf`, `release`, `task`, `trainer`
  - **Allowed types:** `feat`, `fix`, `refactor`, `chore`, `test`
  - Breaking changes: prepend `[BREAKING]` — e.g. `[BREAKING][parallel, model] feat: dynamic batching`

### Test

> Validation results (training curves, eval metrics) for changes not covered by CI.

### API and Usage Example

> Show API changes and usage examples if applicable.

### Design & Code Changes

> High-level design description and specific change list.

### Checklist Before Submitting

- Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md)
- Applied pre-commit checks
- Added/updated documentation
- If `tasks/` training scripts were moved or renamed: updated `docs/` examples and verified `python3 scripts/ci/check_doc_task_paths.py` passes (also enforced by the **Check doc task paths** CI workflow)
- Added tests to CI workflow (or explained why not feasible)

## Summary by CodeRabbit

- **New Features**
  - Added optional FP8 blockwise quantization-aware training for DeepSeek V4.
  - Added FP4 and FP8 fake quantization for activations, weights, KV data, and expert layers.
  - Added support across GPU and NPU execution paths.
  - Added asynchronous activation offload configuration.
  - Added MLU device and MoE implementation support.

- **Improvements**
  - Quantization now supports explicit dequantization without modifying inputs.
  - Improved scale-format and expert data type handling.
  - Added validation for unsupported shapes, formats, input types, and hardware.
  - FP8 QAT requires NVIDIA SM90+ CUDA hardware.

### CODE REVIEW

**Comment by FoolPlayer** [comment]
@coderabbitai full review

File: veomni/models/transformers/deepseek_v4/checkpoint_tensor_converter.py [hunk 1]
```diff
@@ -382,6 +382,8 @@ def export_weights(self, model: torch.nn.Module) -> Generator[tuple[str, torch.T
         config = model.config
         fqn_to_index_mapping = model._veomni_fqn_to_index_mapping
         expert_dtype = config.expert_dtype
+        scale_fmt = "ue8m0"
+        scale_dtype = torch.float8_e8m0fnu if expert_dtype == "fp4" else torch.float32
```
Comment (FoolPlayer) [inline veomni/models/transformers/deepseek_v4/checkpoint_tensor_converter.py:386]: `scale_fmt` goes from a conditional to an unconditional `"ue8m0"` here — just want to confirm this is intentional.

Before this PR it was branched on `expert_dtype` inside the loop:

```python
scale_fmt, scale_dtype = (
    ("ue8m0", torch.float8_e8m0fnu) if expert_dtype == "fp4" else (None, torch.float32)
)
```

So for `expert_dtype != "fp4"` the scale used to be the unrounded fp32 value (`amax / 448`). With `"ue8m0"` every scale is now rounded up to a power of two, which costs one E4M3 mantissa bit.

Since `export_weights` is the generic DSV4 export path and is not gated on QAT, this changes the exported FP8 weights bit-for-bit for **every** DSV4 checkpoint, including runs that never enable QAT.

For QAT itself the change is required — the training-side fake quant defaults to `DEFAULT_SCALE_FMT = "ue8m0"` (`veomni/ops/qat/fp8_blockwise.py`), and if export doesn't match, the quantization simulated during training won't match what actually lands on disk.

So the only open question is how to classify it: was the old `None` branch a bug (in which case this PR fixes it in passing — worth a line in the PR description so downstream notices the numerical change), or was it intentional for non-fp4 checkpoints (in which case this probably needs to be gated on the QAT switch)?

Comment (wuxibin89) [reply veomni/models/transformers/deepseek_v4/checkpoint_tensor_converter.py:386]: Yes, this fix is non-related to QAT, just a bug in main branch. DeepSeek V4 always use `E8M0` scale format regardless of the scale storage format:
- DeepSeek-V4-Flash: stores `scale` in torch.float8_e8m0fnu.
- DeepSeek-V4-Flash-Base: stores `scale` in torch.float32, but still need to round to pow of two.

**Review by FoolPlayer — APPROVED** [review]
